### PR TITLE
New RGB Lighting effect: Twinkle

### PR DESCRIPTION
**Pull request by spidey3**
## Description

This change adds a new RGB Lighting effect: _Twinkle_.

In this mode, underglow LEDs are selected pseudo-randomly to get brighter, and then get darker, over a set period. The result is a twinkling effect.

### Configuration

- There are six sub-modes: 
    0: Monochromatic slow
    1: Monochromatic medium
    2: Monochromatic fast
    3: Multi-color slow
    4: Multi-color medium
    5: Multi-color fast
- Saturation and value (brightness) can be set by the user in all sub-modes.
- In monochromatic sub-modes, hue can be set by the user.
- In multi-color modes, hue is selected pseudo-randomly for each LED.

If you have Velocikey enabled, it will affect the animation speed.

### Implementation

I have tried to keep the increase in code size due to this animation as low as possible, but it does cause an increase in firmware size of about 800 bytes.

## Types of Changes

- [x] Core
- [ ] Bugfix
- [x] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: quantum/rgblight.c [hunk 1]
```diff
@@ -565,7 +566,7 @@ void rgblight_sethsv_at(uint8_t hue, uint8_t sat, uint8_t val, uint8_t index) {
     rgblight_setrgb_at(tmp_led.r, tmp_led.g, tmp_led.b, index);
 }
 
-#if defined(RGBLIGHT_EFFECT_BREATHING) || defined(RGBLIGHT_EFFECT_RAINBOW_MOOD) || defined(RGBLIGHT_EFFECT_RAINBOW_SWIRL) || defined(RGBLIGHT_EFFECT_SNAKE) || defined(RGBLIGHT_EFFECT_KNIGHT)
+#if defined(RGBLIGHT_EFFECT_BREATHING) || defined(RGBLIGHT_EFFECT_RAINBOW_MOOD) || defined(RGBLIGHT_EFFECT_RAINBOW_SWIRL) || defined(RGBLIGHT_EFFECT_SNAKE) || defined(RGBLIGHT_EFFECT_KNIGHT) || defined(RGBLIGHT_EFFECT_TWINKLE)
```
Comment (mtei) [inline quantum/rgblight.c:569]: I think it's better to do the following to make it easier to read.

```diff
diff --git a/quantum/rgblight.c b/quantum/rgblight.c
index 8facd4b09..a24fae5cf 100644
--- a/quantum/rgblight.c
+++ b/quantum/rgblight.c
@@ -566,7 +566,14 @@ void rgblight_sethsv_at(uint8_t hue, uint8_t sat, uint8_t val, uint8_t index) {
     rgblight_setrgb_at(tmp_led.r, tmp_led.g, tmp_led.b, index);
 }
 
-#if defined(RGBLIGHT_EFFECT_BREATHING) || defined(RGBLIGHT_EFFECT_RAINBOW_MOOD) || defined(RGBLIGHT_EFFECT_RAINBOW_SWIRL) || defined(RGBLIGHT_EFFECT_SNAKE) || defined(RGBLIGHT_EFFECT_KNIGHT) || defined(RGBLIGHT_EFFECT_TWINKLE)
+// clang-format off
+
+#if  defined(RGBLIGHT_EFFECT_BREATHING)     \
+  || defined(RGBLIGHT_EFFECT_RAINBOW_MOOD)  \
+  || defined(RGBLIGHT_EFFECT_RAINBOW_SWIRL) \
+  || defined(RGBLIGHT_EFFECT_SNAKE)         \
+  || defined(RGBLIGHT_EFFECT_KNIGHT)        \
+  || defined(RGBLIGHT_EFFECT_TWINKLE)
 
 static uint8_t get_interval_time(const uint8_t *default_interval_address, uint8_t velocikey_min, uint8_t velocikey_max) {
     return
@@ -578,6 +585,8 @@ static uint8_t get_interval_time(const uint8_t *default_interval_address, uint8_
 
 #endif
 
+// clang-format on
+
 void rgblight_setrgb_range(uint8_t r, uint8_t g, uint8_t b, uint8_t start, uint8_t end) {
     if (!rgblight_config.enable || start < 0 || start >= end || end > RGBLED_NUM) {
         return;
```

**Comment by spidey3** [comment]
> The generated code size seems to be a bit larger than the other effects.  I'd like to make it a little smaller, but it seems impossible.

Yes - I put some effort into reducing the code size, and found that this was about the best I could do. Interestingly, using lib8tion's random8() instead of rand() was actually worse, not better...

File: quantum/rgblight.c (hunk 1 above)
Comment (spidey3) [reply quantum/rgblight.c:569]: I thought about that, but preferred to keep my changes to just the functionality.  
Next time I touch this file I will do some formatting cleanup...

**Comment by spidey3** [comment]
gentle ping

**Comment by drashna** [comment]
Thanks!

**Pull request merged**